test/test_iterdatapipe.py

```diff
@@ -370,7 +379,13 @@ def test_header_iterdatapipe(self) -> None:
                 str(wa[0].message), r"length of this HeaderIterDataPipe is inferred to be equal to its limit"
             )
 
+        # __len__ Test: raises TypeError if source doesn't have length and limit is set to None
+        header_dp = source_dp_NoLen.header(None)
+        with self.assertRaises(TypeError):
```

Comment:
TypeError and ValueError are very general and theoretically if we didn't know the implementation, it could also happen because of different reasons. Therefore it is good practice to also check the error message to narrow down the error (e.g. using https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaisesRegex).